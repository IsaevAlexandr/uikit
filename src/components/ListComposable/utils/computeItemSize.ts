import {modToHeight} from '../constants';
import type {ListItemBaseData, ListSizeTypes} from '../types';

export const computeItemSize = (size: ListSizeTypes, item?: ListItemBaseData) => {
    return modToHeight[size][item ? Number(Boolean(item.subtitle)) : 0];
};
