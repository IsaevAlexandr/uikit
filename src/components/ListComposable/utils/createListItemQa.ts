export const createListItemQa = (itemId: string, listId?: string) =>
    listId ? `${listId}-${itemId}` : `${itemId}`;
