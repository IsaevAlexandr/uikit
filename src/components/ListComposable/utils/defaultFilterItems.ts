import type {ListItemType} from '../types';

export function defaultFilterItems<T>(
    items: ListItemType<T>[],
    filterFn: (data: T) => boolean,
): ListItemType<T>[] {
    console.time('defaultFilterItems');
    const getChildren = (result: ListItemType<T>[], item: ListItemType<T>) => {
        if (item.children) {
            const children = item.children.reduce(getChildren, []);

            if (children.length) {
                result.push({data: item.data, children});
            } else if (filterFn(item.data)) {
                result.push({data: item.data, children: []});
            }
        } else if (filterFn(item.data)) {
            result.push(item);
        }

        return result;
    };

    const res = items.reduce<ListItemType<T>[]>(getChildren, []);
    console.timeEnd('defaultFilterItems');
    return res;
}
