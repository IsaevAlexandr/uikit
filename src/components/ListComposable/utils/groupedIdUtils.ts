import {GROUPED_ID_SEPARATOR} from '../constants';
import type {GroupedId} from '../types';

export const getGroupedId = (index: string | number, parentId?: string): GroupedId =>
    parentId ? `${parentId}${GROUPED_ID_SEPARATOR}${index}` : `${index}`;

export const parseGroupedId = (parentId: GroupedId): string[] =>
    parentId.split(GROUPED_ID_SEPARATOR);
