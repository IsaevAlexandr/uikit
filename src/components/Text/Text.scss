@import '../variables';
@import '../../../styles/mixins';

$block: '.#{$ns}text';

#{$block} {
    &_type_display1 {
        @include text-display-1();
    }
    &_type_display2 {
        @include text-display-2();
    }
    &_type_display3 {
        @include text-display-3();
    }
    &_type_display4 {
        @include text-display-4();
    }
    &_type_code1 {
        @include text-code-1();
    }
    &_type_code2 {
        @include text-code-2();
    }
    &_type_code3 {
        @include text-code-3();
    }
    &_type_codeInline1 {
        @include text-code-inline-1();
    }
    &_type_codeInline2 {
        @include text-code-inline-2();
    }
    &_type_codeInline3 {
        @include text-code-inline-3();
    }
    &_type_body1 {
        @include text-body-1();
    }
    &_type_body2 {
        @include text-body-2();
    }
    &_type_body3 {
        @include text-body-3();
    }
    &_type_bodyShort {
        @include text-body-short();
    }
    &_type_caption1 {
        @include text-caption-1();
    }
    &_type_caption2 {
        @include text-caption-2();
    }
    &_type_header1 {
        @include text-header-1();
    }
    &_type_header2 {
        @include text-header-2();
    }
    &_type_subheader1 {
        @include text-subheader-1();
    }
    &_type_subheader2 {
        @include text-subheader-2();
    }
    &_type_subheader3 {
        @include text-subheader-3();
    }
    &_ellipsis {
        @include overflow-ellipsis();
    }
}
