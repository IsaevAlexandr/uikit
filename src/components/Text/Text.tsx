import React from 'react';
import {text, TextBase} from './text/text';
import {colorText, ColorTextBase} from './colorText/colorText';

export interface TextProps extends TextBase, ColorTextBase {
    /**
     * Ability to override default html tag
     */
    as?: keyof JSX.IntrinsicElements;
    style?: React.CSSProperties;
    className?: string;
    children?: React.ReactNode;
}

/**
 * A component for working with typography.
 *
 * Storybook: https://preview.yandexcloud.dev/uikit/?path=/docs/components-text--default
 *
 * **Hint:** Hover on props in your editor to read jsdoc
 *
 * Provides a convenient API for working with mixins of typography and text colors. Just point at the prop in you favorite code editor and read the accompanying documentation via `jsdoc` on where to apply this or that font or color.
 *
 * ```jsx
 * import {Text} from '@yandex-cloud/uikit';
 *
 * <Text typography="body-1" color="inherit" elipsis>some test</Text>
 * ```
 *
 * You can also use a more flexible way of setting the style. "Uikit" alse provide `text` uility function.
 *
 *```jsx
 * import {text} from '@yandex-cloud/uikit';
 *
 * // textStyles = 'text text_typography_display-1 some-class-name'
 * const textStyles = text({typography: 'display-1'}, 'some-class-name'));
 *
 * <span className={textStyles}>some text</span>
 * ```
 */
export const Text: React.FC<TextProps> = ({
    as: Tag = 'span',
    children,
    typography,
    className,
    ellipsis,
    color,
    ...rest
}) => {
    return (
        <Tag
            className={text(
                {typography, ellipsis},
                color ? colorText({color}, className) : className,
            )}
            {...rest}
        >
            {children}
        </Tag>
    );
};
