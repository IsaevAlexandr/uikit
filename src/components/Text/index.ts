export * from './Text';
export {TYPOGRAPHY_VARIANTS, text} from './text/text';
export {TEXT_COLORS, colorText} from './colorText/colorText';
