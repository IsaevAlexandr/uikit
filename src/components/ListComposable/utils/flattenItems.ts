import type {ListItemId, ListItemType} from '../types';

import {getGroupedId} from './groupedIdUtils';

export function flattenItems<T>(
    items: ListItemType<T>[],
    groupsExpandedState: Record<ListItemId, boolean> = {},
): ListItemId[] {
    console.time('flattenItems');

    const getNestedIds = (
        order: string[],
        item: ListItemType<T>,
        index: number,
        parentId?: string,
    ) => {
        const groupedId = getGroupedId(index, parentId);
        const id = item.id || groupedId;

        order.push(id);

        if (item.children) {
            // remove from order collapsed groups
            if (!(id in groupsExpandedState && !groupsExpandedState[id])) {
                order.push(
                    ...item.children.reduce<string[]>(
                        (acc, item, idx) => getNestedIds(acc, item, idx, groupedId),
                        [],
                    ),
                );
            }
        }

        return order;
    };

    const result = items.reduce<string[]>((acc, item, index) => getNestedIds(acc, item, index), []);
    console.timeEnd('flattenItems');
    return result;
}
