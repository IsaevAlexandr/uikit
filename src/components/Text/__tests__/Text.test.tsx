import React from 'react';

import {render} from '../../../../test-utils/utils';
import {Button} from '../../Button';
import {Text} from '../Text';

describe('Text', () => {
    describe('should return expected html', () => {
        test('if no props passed', () => {
            const {container} = render(<Text>Hello World!</Text>);

            expect(container).toMatchSnapshot();
        });

        test('if qa attr passed and variant changed', () => {
            const {container} = render(
                <Text qa="test" variant={'caption-1'}>
                    Hello World!
                </Text>,
            );

            expect(container).toMatchSnapshot();
        });

        test('if html attributes passed and changed default html tag and with typed ref', () => {
            const ComponentWithRef = () => {
                const ref = React.useRef<HTMLLabelElement>(null);

                return (
                    <Text as="label" htmlFor="some-id" ref={ref}>
                        Hello World!
                    </Text>
                );
            };

            const {container} = render(<ComponentWithRef />);

            expect(container).toMatchSnapshot();
        });

        test('if passed props what converted to classNames and styles', () => {
            const {container} = render(
                <Text
                    style={{width: 200}}
                    ellipsisLines={2}
                    ellipsis={true}
                    whiteSpace="break-spaces"
                    wordBreak="break-word"
                >
                    Hello World!
                </Text>,
            );

            expect(container).toMatchSnapshot();
        });

        test('with another component substitution', () => {
            const {container} = render(
                <Text as={Button} size={'m'} view="action">
                    Hello World!
                </Text>,
            );

            expect(container).toMatchSnapshot();
        });
    });
});
