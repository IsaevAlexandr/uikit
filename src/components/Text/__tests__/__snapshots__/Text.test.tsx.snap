// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Text should return expected html if html attributes passed and changed default html tag and with typed ref 1`] = `
<div>
  <label
    class="g-text g-text_variant_body-1"
    for="some-id"
  >
    Hello World!
  </label>
</div>
`;

exports[`Text should return expected html if no props passed 1`] = `
<div>
  <span
    class="g-text g-text_variant_body-1"
  >
    Hello World!
  </span>
</div>
`;

exports[`Text should return expected html if passed props what converted to classNames and styles 1`] = `
<div>
  <span
    class="g-text g-text_variant_body-1 g-text_ellipsis g-text_ws_break-spaces g-text_wb_break-word g-text_ellipsis-lines"
    style="width: 200px;"
  >
    Hello World!
  </span>
</div>
`;

exports[`Text should return expected html if qa attr passed and variant changed 1`] = `
<div>
  <span
    class="g-text g-text_variant_caption-1"
    data-qa="test"
  >
    Hello World!
  </span>
</div>
`;

exports[`Text should return expected html with another component substitution 1`] = `
<div>
  <button
    class="g-button g-button_view_action g-button_size_m g-button_pin_round-round g-text g-text_variant_body-1"
    type="button"
  >
    <span
      class="g-button__text"
    >
      Hello World!
    </span>
  </button>
</div>
`;
